### ISSUE TITLE
clang-format slow on big files compared to vim

### ISSUE THREAD

**Issue by oni-link** [neovim/neovim#1410]
clang-format in Neovim is noticeable slower when used on big files compared to Vim.

Testing on src/nvim/eval.c, formating the whole file:

``` vim
:let start=reltime()| execute "1,$ pyf /path/to/clang-format.py" |echo 'system() = ' . reltimestr(reltime(start))
```

Neovim: ~12.9s
Vim      : ~0.6s

Where does this slowdown come from?

Neovim version: 4cc26fba249361cccd791a76fc35c46d851096a5
Build command: rm -rf build/ && make cmake CFLAGS='-Wno-error -DNDEBUG -O3 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1' && make -j5

clang-format version 3.6.0 (http://llvm.org/git/clang.git f65002072fe308f93a0ef04054b56218b00c2a34) (http://llvm.org/git/llvm.git 84dc1b5c92becc0c7f87d0e16afa57e0bcf84dd4)

**Comment by justinmk** [comment]
@oni-link Is this still an issue?

**Comment by oni-link** [comment]
Yes, it is. New measurements for nvim with 'Debug' and 'Release' mode give `14.0` seconds, for vim I get `0.7` seconds.

**Comment by oni-link** [comment]
Measuring the time for formating all of `eval.c` directly in the [clang-format.py](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/clang-format.py), I found that the following loop accounts for nearly the total time of `14s`:
```py
    sequence = difflib.SequenceMatcher(None, vim.current.buffer, lines)
    for op in reversed(sequence.get_opcodes()):
      if op[0] is not 'equal':
        vim.current.buffer[op[1]:op[2]] = lines[op[3]:op[4]]
```
Every use of `vim.current.buffer` takes approx. `10^-4s` ( for vim it is `10^-5s...10^-7s`) and accounts for `0.1s` ( ~`1000` times called).
Looks like the use of `vim.current.buffer` in `sequence = difflib.SequenceMatcher(None, vim.current.buffer, lines)` is the problem here.
To speed up the computation of `op`, I replaced `vim.current.buffer` with the buffer content in `difflib.Sequence(...)`:
```diff
diff --git a/tools/clang-format/clang-format.py b/tools/clang-format/clang-format.py
index 5cb41fc..4c2dd75 100644
--- a/tools/clang-format/clang-format.py
+++ b/tools/clang-format/clang-format.py
@@ -49,8 +49,8 @@ if vim.eval('exists("g:clang_format_fallback_style")') == "1":
 
 def main():
   # Get the current text.
-  buf = vim.current.buffer
-  text = '\n'.join(buf)
+  buf_lines = vim.current.buffer[:]
+  text = '\n'.join(buf_lines)
 
   # Determine range to format.
   if vim.eval('exists("l:lines")') == '1':
@@ -95,7 +95,7 @@ def main():
     lines = stdout.split('\n')
     output = json.loads(lines[0])
     lines = lines[1:]
-    sequence = difflib.SequenceMatcher(None, vim.current.buffer, lines)
+    sequence = difflib.SequenceMatcher(None, buf_lines, lines)
     for op in reversed(sequence.get_opcodes()):
       if op[0] is not 'equal':
         vim.current.buffer[op[1]:op[2]] = lines[op[3]:op[4]]
```
That brought the total time down from `14.55s` to `0.807s`. That is comparable with vims `0.703s`.

Do we have a problem with slow API functions? Is this fixable in neovim?

**Comment by tarruda** [comment]
> Do we have a problem with slow API functions? Is this fixable in neovim?

TL;DR It might be possible to fix in the python host.

It seems the difference is caused by the overhead of IPC(serialization/copy/deserialization), and if so, the clang script has hit a limitation of Neovim emulation of vim's python API: A plugin can't send IPC requests in a loop as if it was calling a function in the same process.

I once had the same problem with another plugin(I think it was YCM) that used `:pydo` to process lines, my solution was to cache the lines in the python process to mitigate the IPC overhead: https://github.com/neovim/python-client/blob/4f9d930ebb110eeeacc7129366bfde3c86c72448/neovim/plugin/script_host.py#L108-L111

The caching approach is relatively safe because when nvim is sending requests to a plugin host, it blocks until the initial request returns and only handles requests from the host which received the request, so it is very unlikely that the buffer will have its contents changed.

I say unlikely because it is still possible for the plugin to execute a command that is implemented by another plugin host, which could modify the contents before the request to the python host is finished(Neovim would temporarily switch to handling requests from the new host), causing the cache to be outdated.

A way to make the caching approach more robust would be to have neovim send an "invalidate" notification to all hosts that are currently processing a request but are not the active plugin host, such notification could be used to invalidate the host caches.

**Comment by bfredl** [comment]
> the overhead of IPC(serialization/copy/deserialization), 

There is also the cost of roundtrips (context switches in this case). That `SequenceMatcher` would have two context-switches not only per line, but per _line_ _access_, which probably dwarfs the cost of just copying the data.

> send an "invalidate" notification to all hosts.

Also for `vim.current` it seems reasonable to use update notifications (in this case not only "invalidation", but the actual value) instead of two context switches per `vim.current.buffer`. In an async context we are, well, async anyway and it is enough to be "eventually consistent" and for sync requests we need to make sure to send the notification to that client before the request itself, and if the value was changed in a nested request, before the reply is sent to the client. For a naive plugin that does `vim.current.buffer[i]` this will already halve the cost.

**Comment by amandeepgautam** [comment]
Any workarounds yet? I am still seeing this on Ubuntu 18, with NVIM v0.4.4.

**Comment by justinmk** [comment]
@amandeepgautam [clang_format.py](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/clang-format.py) is pretty simple, so a reasonable workaround would be to rewrite it in Lua (which runs in-process in Nvim, unless Pythin).

**Comment by amandeepgautam** [comment]
@justinmk Would plugins be any different? ex: https://github.com/andreyorst/SimpleClangFormat.vim or https://github.com/sbdchd/neoformat.

**Comment by justinmk** [comment]
> Would plugins be any different? ex: https://github.com/andreyorst/SimpleClangFormat.vim or https://github.com/sbdchd/neoformat or https://vimawesome.com/plugin/vim-clang-format.

this issue is specifically about clang_format.py which is calling back-and-forth many times between nvim and python.  If those plugins use clang_format.py then they will have the same issue...

**Comment by OscarCreator** [comment]
Is this really an issue still?

I ran the command on eval.c with the latest [clang-format.py](https://github.com/llvm/llvm-project/blob/main/clang/tools/clang-format/clang-format.py) and I get ~4s on eval.c in both neovim and vim. And after formated once i get ~1s
